File: serialInterfaceScripts/MATLAB/OS_moveSpeed.m
function varargout=OS_moveSpeed(mSpeed)
% Set or read axes speed parameters on OpenStage
%
% function speeds=OS_moveSpeed(mSpeed)
%
%
% Purpose
% Set or read the speed parameter of each axis. If no
% inputs or outputs are provided, prints values to screen.
% If an output is provided, values are assigned to this as a
% vector. If mSpeed is provided, the function sets the 
% speed parameters on the stage. Parameters relate to
% Go To serial motions and right-button motions only. 
%
%
% Inputs
% mSpeed - [optional] a vector defining the speed in microns 
%          per second for each axis. Axes ordered X,Y,Z by
%          default. All must be provided.
%
% Outputs
% speeds - [optional] a vector listing the speeds of each 
%          axis in microns per second. 
%
%
% Examples
% OS_moveSpeed     %prints axes max speeds to screen
% sp=OS_moveSpeed; %returns axes max speeds as variable sp
% OS_moveSpeed([500,500,1000]) %set speeds in X, Y, and Z
%
%
% Rob Campbell - CSHL, August 2013



global OS;
if isempty(OS), OS=connectOpenStage; end

%flush buffer if needed
if get(OS,'BytesAvailable')>0
    fread(OS);
end


%First read from controller. This is done regardles of what the
%user asked for in order to get the number of axes.
fwrite(OS,'vr')
str=fscanf(OS);

numAxes=length(strfind(str,','))+1;
scanString=repmat('%f,',1,numAxes);
scanString(end)=[];

speeds=textscan(str,[scanString,'$']);
speeds=cell2mat(speeds);


%Handle parameter change request
if nargin==1
  if any(mSpeed<0)
      fprintf('one or more mSpeed values out of range\n')
      return
  end    
  if length(mSpeed) ~= numAxes
      fprintf('Expected %d axes but got %d\n',...
          numAxes, length(mSpeed))
      return
  end
  
  scanString=repmat('%d,',1,numAxes);
  scanString(end)=[];
  str=sprintf(['vs',scanString,'$'],mSpeed);
  fwrite(OS,str)
  
end


%Handle outputs
if nargout==0 & nargin==0
    fprintf('%s\n',str(1:end-1))
end

if nargout>0
    varargout{1}=speeds;
end

File: serialInterfaceScripts/Python/OS_moveSpeed.py
# Purpose
# Set or read the speed parameter of each axis. If no
# inputs are provided, prints values to screen. If three inputs 
# are provided, the function uses these to set the speed 
# parameters on the stage. Parameters relate to Go To serial 
# motions and right-button motions only. 
#
#
# Inputs
# Three numbers defining the speed in microns per second for 
# Axes ordered X,Y,Z and all must be provided.
#
# Examples
# OS_moveSpeed     #prints axes speeds to screen
# OS_moveAccel 500 500 1000 #set speeds in X, Y, and Z
#
#
# Rob Campbell - CSHL, December 2013



import sys
from OpenStageSerial import ser, readAndPrintBuffer, readBuffer

import re

ser.flushInput() #flush the buffer, to be safe


#First read from controller. This is done regardles of what the
#user asked for in order to get the number of axes.
ser.write('vr')
buf=readBuffer()
numAxes=len([m.start() for m in re.finditer(',', buf)]) + 1


#Either report parameters to the screen or change them
if len(sys.argv)<2:
  print(buf)
  sys.exit(0)


if len(sys.argv)<numAxes+1:
  sys.stderr.write("Must provide all axes\n")
  sys.exit(0)



#Build string to send
strToSend='vs'

for arg in sys.argv[1:4]:
  arg = int(arg)

  if arg<0:
    sys.stderr.write("Value out of range\n")
    sys.exit(0)

  strToSend += str(arg) + ','

strToSend=strToSend[:-1] + '$'


#Send string
ser.write(strToSend)
ser.close()